### ISSUE TITLE
hydra hashed mirror / tarballs.nixos.org ignores the name & version

### ISSUE THREAD

**Issue by darklajid** [NixOS/nixpkgs#2565]
Hey.

So I'm new. Please don't hit me, or use a cluebat at least.
I had a long discussion on IRC with mm_freak and viric.

The beginning: I bumped the version of a couple of expressions, using a (stupid?) workflow like this:

- Grab the .nix from a recent git checkout, branch master, and copy it to a local temporary folder, naming it $pkg.nix.

- Change the first line to the one for out of tree packages in the manual [1], ie. 
  with import <nixpkgs> {}; # bring all of Nixpkgs into scope

- Bump the version to the one I want (example: tmux from 1.8 to 1.9)

- Run nix-build $pkg.nix

At this point I expect nix-build to fail, because I was clueless at the beginning and now I'm a lazy bastard. Clueless before: I didn't know about nix-prefetch-url. Lazy now: I still like the verification of nix-build failing with a wrong checksum error and just copy/paste the correct one from there.

The problem:

So while that process worked fine three times in a row, last time roughly an hour ago with tmux (1.8 -> 1.9), now I tried to bump taskwarrior.

Copied the file, changed the first line to make it possible to test it out of the tree, bumped the version from 2.2.0 to 2.3.0 and ran nix-build taskwarrior.nix.

The resulting expression:

```
with import <nixpkgs> {}; # bring all of Nixpkgs into scope  
 
stdenv.mkDerivation rec {
  name = "taskwarrior-${version}";
  version = "2.3.0";
 
  enableParallelBuilding = true;
 
  src = fetchurl {
    url = "http://www.taskwarrior.org/download/task-${version}.tar.gz";
    sha256 = "057fh50qp9bd5s08rw51iybpamn55v5nhn3s6ds89g76hp95vqir";
  };
 
  nativeBuildInputs = [ cmake ];
 
  meta = {
    description = "GTD (getting things done) implementation";
    homepage = http://taskwarrior.org;
    license = stdenv.lib.licenses.mit;
    maintainers = [stdenv.lib.maintainers.marcweber];
    platforms = stdenv.lib.platforms.linux;
  };
}
```

That's the current version in master, with just the version number changed. The hash is wrong, is still the old one (see process above).

Partial output of nix-build:

```
these derivations will be built:
  /nix/store/74mklcaldiihksyhzdcg0vmskb3czkbc-taskwarrior-2.3.0.drv
  /nix/store/py1c2bzx2w8xqkcrwadxvy28914a549b-task-2.3.0.tar.gz.drv
these paths will be fetched (4.21 MiB download, 37.78 MiB unpacked):
  /nix/store/fll4gmsmfy8sahbrzvvqhii55v0k141n-cmake-2.8.12.2
fetching path `/nix/store/fll4gmsmfy8sahbrzvvqhii55v0k141n-cmake-2.8.12.2'...
building path(s) `/nix/store/l6rhd5iiq9xrkdrrwflqv3s8c39szc05-task-2.3.0.tar.gz'
 
ESC[ptrying http://tarballs.nixos.org/sha256/057fh50qp9bd5s08rw51iybpamn55v5nhn3s6ds89g76hp95vqir
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
^M  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
*** Downloading ‘http://cache.nixos.org/nar/1cahc9il9gk9f3fwc481hxnxmll7jvgzlxa5jqa5a0j0pzrl8gzh.nar.xz’ to ‘/nix/store/fll4gmsmfy8sahbrzvvqhii55v0k141n-cmake-2.8.12.2’...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
^M  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0^M 19 4313k   19  861k    0     0  1581k      0  0:00:02 --:--:--  0:00:02 1581k^M 73 1162k   73  857k    0     0  1492k      0 --:--:-- --:--:-- --:--:-- 1491k^M100 1162k  100 1162k    0     0  1605k      0 --:--:-- --:--:-- --:--:-- 1604k
ESC[qESC[q^M100 4313k  100 4313k    0     0  3029k      0  0:00:01  0:00:01 --:--:-- 3031k
 
building path(s) `/nix/store/z3ywwwmp8ql42xi166zs3nf3jfc8hkbq-taskwarrior-2.3.0'
ESC[pbuilding /nix/store/z3ywwwmp8ql42xi166zs3nf3jfc8hkbq-taskwarrior-2.3.0
ESC[punpacking sources
ESC[3punpacking source archive /nix/store/l6rhd5iiq9xrkdrrwflqv3s8c39szc05-task-2.3.0.tar.gz
ESC[qsource root is task-2.2.0
ESC[qESC[ppatching sources
```

Note that

- it grabbed the tarball by its hash only

- it builds 2.2.0 (successfully)

- result/bin/task --version is now reporting 2.2.0

I kind of understand by now that I shouldn't work like this, but the unpredictable nature here (works three times in a row, failed now and might work with the very  same expression again if the mirror happens to drop the tarball) and explanations like "Don't do that" or (quoting vivic) "it's a matter of complex details, what that mirror is mirroring and what not" indicate to me that this isn't optimal..

vivic suggested that this might be fixed if the version/name would be taken into account as part of the key for tarballs.nixos.org.

Currently the hash alone is used to grab the tarball. That means that a package with an old hash, updated version/fetch url might silently 'work' and build something unexpected.

Therefor I propose to include the version/pkg name into the key for cached tarballs. That way the cache works just fine (after being primed again), just like now. It won't give an unexpected hit in cases like mine though.

Currently the dependency between fetch url and hash is messed up imho: Fetch url should be the master source for the tarball, hash the verification. This cache service turns things on its head by assuming the hash is the master key and ignoring the fetch url (or name or version).

Bottom line: Please consider changing checks for

http://tarballs.nixos.org/sha256/057fh50qp9bd5s08rw51iybpamn55v5nhn3s6ds89g76hp95vqir

to something similar to what nix-store uses

http://tarballs.nixos.org/sha256/057fh50qp9bd5s08rw51iybpamn55v5nhn3s6ds89g76hp95vqir-taskwarrior-2.3.0

1: http://nixos.org/nixos/manual/#sec-custom-packages

**Comment by Mathnerd314** [comment]
Recent versions of ``pkgs.nixUnstable`` include a new ``nix-prefetch-url`` command; unlike the old one, it supports ``-A``. Running ``nix-prefetch-url taskwarrior.nix -A src`` will fetch the new tarball and print out the new hash, for easy pasting into the file.

Is this is sufficient?

**Comment by vcunat** [comment]
I'm personally not convinced this use case is a strong enough reason to cause an incompatible change in the way how our hashed mirrors work (or at least not enough to do it myself).

**Comment by grahamc** [comment]
I move to close this issue.

**Comment by joachifm** [comment]
Agreed
